Slim/Formats/Playlists/CUE.pm

```diff
@@ -60,144 +122,590 @@ sub parse {
+		# instead of storing them as album / previous Track related as before.
+		#
+		# All checks on values that imply a lookup to other has been delayed at the
+		# step after the line loop. 
+		#
+		# Here some basic validation check on single commands when:
+		# 1. relative position has a meaning (i.e TRACK, INDEX, REM END, FILE)
+		# 2. special syntax validation is needed (i.e. REPLAYGAIN, COMPILATION).
+		#
+		# Most of what was done here before has been moved after the line loop.
+		# 
+		my($command,$value)= _getCommandFromLine($line);
+
+		if (!defined $command or !defined $value){
+
+			#No commads in line, skipping;
+
+			my $msg = {};
+			$msg->{'message'}	= 'No command in line: Skipping';
+			$msg->{'line'}		= $line;
+			$msg->{'command'}	= $command;
+			$msg->{'value'}		= $value;
+
+			main::DEBUGLOG && $log->is_debug && $log->debug(Data::Dump::dump($msg));
+
+		} elsif (!_isCommandAccepted($command)){
+
+			#Command refused;
+
+			my $msg = {};
+			$msg->{'message'}	= 'Command refused';
+			$msg->{'line'}		= $line;
+			$msg->{'command'}	= $command;
+			$msg->{'value'}		= $value;
+
+			main::DEBUGLOG && $log->is_debug && $log->debug(Data::Dump::dump($msg));
+
+		} elsif ($command eq 'TRACK'){
+
+			$inAlbum=0;
+
+			#Skipping non audio tracks.
+			if ($value =~ /^(\d+)\s+AUDIO/i) {
+
+				$currtrack = int($1);
+
+			} elsif ($value =~ /^(\d+)\s+.*/i){
+
+				$currtrack=undef;
+			}
 
-			if (defined $currtrack) {
+		} elsif ($command eq 'INDEX'){
 
-				$tracks->{$currtrack}->{'TITLE'} = $1;
+			if (!defined $currtrack){
+				#Ignored
+				main::DEBUGLOG && $log->debug("Index found for missing Track");
 
-			} else {
+			}elsif ($value =~ /^00\s+(\d+):(\d+):(\d+)/i) {
 
-				$cuesheet->{'ALBUM'} = $1;
+				$tracks->{$currtrack}->{'PREGAP'} = ($1 * 60) + $2 + ($3 / 75);
+
+			}elsif ($value =~ /^01\s+(\d+):(\d+):(\d+)/i){
+
+				$tracks->{$currtrack}->{'START'} = ($1 * 60) + $2 + ($3 / 75);
 			}
 
-		} elsif ($line =~ /^PERFORMER\s+\"(.*)\"/i) {
+		}elsif ($command eq 'REM') {
 
-			$cuesheet->{'ARTIST'} = $1;
-			$cuesheet->{'ALBUMARTIST'} = $1;
-			
-		} elsif ($line =~ /^(?:REM\s+)?(YEAR|GENRE|COMMENT|ARTISTSORT|ALBUMARTISTSORT|ALBUMSORT|COMPILATION)\s+\"(.*)\"/i) {
+			my($remCommand,$remValue)= _getRemCommandFromLine($value);
 
-			$cuesheet->{uc($1)} = $2;
+			if (!defined $remCommand or !defined $remValue){
 
-		} elsif ($line =~ /^(?:REM\s+)?(?:DISC|DISCNUMBER)\s+\"?([^\"]+)\"?/i) {
+				#No commads in rem, skipping;
 
-			# Some Cue sheets may not bother with quotes around a number
-			$cuesheet->{'DISC'} = $1;
+				my $msg = {};
+				$msg->{'message'}	= 'No commads in rem, skipping';
+				$msg->{'line'}		= $line;
+				$msg->{'command'}	= $command;
+				$msg->{'value'}		= $value;
+				$msg->{'remCommand'}	= $remCommand;
+				$msg->{'remValue'}	= $remValue;
 
-		} elsif ($line =~ /^(?:REM\s+)?(?:DISCC|DISCTOTAL|TOTALDISCS)\s+\"?([^\"]+)\"?/i) {
+				main::DEBUGLOG && $log->is_debug && $log->debug(Data::Dump::dump($msg));
```

Comment:
One more optimization please: we don't want to deal with that hash unless we are actually logging. We can make it inline:

```perl
main::DEBUGLOG && $log->is_debug && $log->debug(Data::Dump::dump({
   message => 'No commands in rem, skipping',
   line          => $line,
   ...
}));
```
Saves us some CPU cycles and bytes when not logging.